### PR TITLE
Limit index length to 191 characters by default, additionally connect HPOS to verify DB tooling.

### LINKED ISSUE #39051: HPOS SQL Code Too Large
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

**Issue Description:**

The wp_wc_orders table creation is not compatible with some shared servers. When trying to create this table (see bottom of post), I get the error, "Specified key was too long; max key length is 1000 bytes."

After contacting my shared server (Nixihost), I'd have to change global database versions or variables, which is not allowed. The only way to allow this query is to upgrade to a dedicated server (an extra $120 a month).

Which keys are over 1000 bytes? Is it possible to modify this SQL query to keep it under the limits?

`CREATE TABLE wp_wc_orders ( id bigint(20) unsigned, status varchar(20) null, currency varchar(10) null, type varchar(20) null, tax_amount decimal(26,8) null, total_amount decimal(26,8) null, customer_id bigint(20) unsigned null, billing_email varchar(320) null, date_created_gmt datetime null, date_updated_gmt datetime null, parent_order_id bigint(20) unsigned null, payment_method varchar(100) null, payment_method_title text null, transaction_id varchar(100) null, ip_address varchar(100) null, user_agent text null, customer_note text null, PRIMARY KEY (id), KEY status (status), KEY date_created (date_created_gmt), KEY customer_id_billing_email (customer_id, billing_email), KEY billing_email (billing_email), KEY type_status (type, status), KEY parent_order_id (parent_order_id), KEY date_updated (date_updated_gmt) ) DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci
`

### PR DESCRIPTION
**Pull request by vedanshujain**
Additionally, connect verify db tooling to order tables when they are enabled.

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR introduces two changes:

1. It limited the index length to `191` characters as we do for other WooCommerce schema. This is needed because at least in the MyISAM engine, the maximum index size can only be 1000 bytes, which is occupied by 192 chars with utf8_mb4 encoding.
2. It links the already existing `verify_db_tool` with the HPOS schema if it's enabled (or if data sync is on). This tools alerts the site admin if a table is missing.

Closes #39051 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Start with a site with posts authoritative and sync disabled. Delete the HPOS tables if they already exist by going to WooCommerce > Status > Tools and running the `Delete the custom orders tables` tool.
2. Add this filter to simulate the error first: 
```
add_filter( 'woocommerce_database_max_index_length', function () {
	return 320; // will cause error on MyISAM but works fine on InnoDB.
} );
```
3. We will now modify the code to create one of the tables with MyISAM which will simulate the error. Edit this file: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php#L2664 so that it now says `) $collate ENGINE = MyISAM;` instead of just `) $collate ;`.
4. Once the file is edited, let's try to create the table. Go to WooCommerce > Settings > Advance > Features and enable the option to keep posts and orders in sync. This will create all tables but the meta table, which will fail with error: `Specified key was too long; max key length is 1000 bytes`.
5. Go to WooCommerce > Status > Tools and run the `Verify base database tables` tool. It will identify the missing table and show this error: `Verifying database... One or more tables are still missing: wp_wc_orders_meta`.
6. Now, remove the filter that we added in Step 2. Go to WooCommerce > Settings > Advance > Features again and enable/disable the sync feature. This will trigger the table creation routine again.
7. Come back to WooCommerce > Status > Tools and run the `Verify base database tables` tool. It should show that all tables are present as expected.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement

#### Comment 

</details>

### CODE REVIEW

**Comment by rrennick** [comment]
> This will create all tables but the meta table, which will fail with error: Specified key was too long; max key length is 1000 bytes.

@vedanshujain Should we update this message to reflect the limit imposed by the filter?

File: plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php [hunk 1]
```diff
@@ -366,6 +366,7 @@ private function handle_data_sync_option_changed( string $feature_id ) {
 
 		if ( ! $this->data_synchronizer->check_orders_table_exists() ) {
 			$this->data_synchronizer->create_database_tables();
+
```
Comment (rrennick) [inline plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php]: This looks accidental.

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php [hunk 2]
```diff
@@ -2584,6 +2584,23 @@ public function get_database_schema() {
 		$operational_data_table_name = $this->get_operational_data_table_name();
 		$meta_table                  = $this->get_meta_table_name();
 
+		/**
```
Comment (rrennick) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php]: Would it be worth adding this filter in its own function in a utility/helper class as we may find more uses for it as we work through some of the other upcoming data storage projects?

**Review by rrennick — COMMENTED** [review]
Code looks good but I didn't test it yet. I left that for after resolving the discussion points.

**Comment by rrennick** [comment]
@vedanshujain Not for this PR but there are couple small things I noticed:

- When the meta table was missing the delete tables tool wasn't listed in the tools even though the other tables were
- The sync went ahead whether or not the meta table was present. It would be better if it was stopped and the merchant had a notification of some kind

**Review by rrennick — APPROVED** [review]
Nice work on this @vedanshujain . It tested great for me.

**Comment by nigeljamesstevenson** [comment]
Thanks a lot for the input on this @rrennick - much appreciated! @vedanshujain, do you want us to create Issues for the [suggestions](https://github.com/woocommerce/woocommerce/pull/39250#issuecomment-1658638929) that Ron has made?

**Comment by coreymckrill** [comment]
@vedanshujain I just noticed that WC is giving a warning that WC Smooth Generator is not compatible with HPOS, even though it has been [declaring compatibility](https://github.com/woocommerce/wc-smooth-generator/blob/5c0e08805825137cd4d4ccf65d413127dede3106/wc-smooth-generator.php#L54-L61) for quite a while now. I used `git bisect` to trace the change to this PR. From the changeset I don't immediately see what could be causing this, but wanted to give a heads up, as this could be affecting other extensions as well.

**Comment by rrennick** [comment]
> I used git bisect to trace the change to this PR. From the changeset I don't immediately see what could be causing this

It's not registered as compatible with the two new features added in this PR.

**Pull request merged**